File: services/mirror-worker/src/replay/order-changed.ts
import axios from 'axios';
import { PoolClient } from 'pg';
import { config } from '../config';
import { log } from '../log';
import { ProdUserLog } from '../db/prod';
import { getStagingApiToken } from '../runtime';
import {
  translateEquipmentId,
  translateProductionOrderId,
} from '../translate';

// Prod payload shape captured from session 55 inspection:
//   {
//     idArea, idSite,
//     idOrder: <string biz key>,
//     stopType: 'pause' | 'finish',
//     timestamp: ISO,
//     idEquipment: <prod>,
//     idEnterprise: <prod>,
//     equipmentSetup: [{ id, position }],
//     shouldCreatePo, shouldOpenNewPo, unitMultiplier,
//     idProductionOrder: <string biz key, same as idOrder>,
//     oldIdProductionOrder: <prod surrogate — needs translation>,
//     productionOrderQuantity: <string>,
//     oldProductionOrderProdFinal
//   }
//
// Translation:
//   idEquipment            -> via packml_topic
//   oldIdProductionOrder   -> via mirror_id_map (entity_type='production_order')
//                             populated by Phase A6 backfill for pre-existing
//                             prod POs and by future order-created-started
//                             replays for any new ones
//
// Output: POST /api/production-orders/stop with the StopProductionOrderDto
// shape (timestamp, stopType, idEnterprise, idEquipment, idProductionOrder,
// productionOrderQuantity). Extra fields the operator sent that don't appear
// in the Stop DTO are ignored on staging — class-validator doesn't reject
// non-whitelisted fields by default.
interface OrderChangedPayload {
  stopType: string;
  timestamp: string;
  idEquipment: number;
  idEnterprise: number;
  oldIdProductionOrder: number;
  productionOrderQuantity: number | string;
}

export async function replayOrderChanged(
  _client: PoolClient,
  row: ProdUserLog,
): Promise<void> {
  const payload = row.payload as OrderChangedPayload;
  if (!payload || typeof payload !== 'object') {
    throw new Error('order-changed payload missing or not an object');
  }
  if (typeof payload.idEquipment !== 'number') {
    throw new Error('order-changed payload missing idEquipment');
  }
  if (typeof payload.oldIdProductionOrder !== 'number') {
    throw new Error('order-changed payload missing oldIdProductionOrder');
  }
  if (!payload.stopType) {
    throw new Error('order-changed payload missing stopType');
  }

  const stagingEquipmentId = await translateEquipmentId(payload.idEquipment);
  const stagingPoId = await translateProductionOrderId(
    BigInt(payload.oldIdProductionOrder),
  );
  if (stagingPoId === undefined) {
    throw new Error(
      `production_order ${payload.oldIdProductionOrder} unmapped (no mirror_id_map row and no nu_production_order business-key match — run backfill-pos for currently-running prod POs)`,
    );
  }

  const body = {
    timestamp: payload.timestamp,
    stopType: payload.stopType,
    idEnterprise: config.stagingEnterpriseId,
    idEquipment: stagingEquipmentId,
    idProductionOrder: Number(stagingPoId),
    productionOrderQuantity: Number(payload.productionOrderQuantity),
  };

  const url = `${config.stagingApi.baseUrl}/api/production-orders/stop`;
  const res = await axios.post(url, body, {
    params: {
      token: getStagingApiToken(),
      idEnterprise: config.stagingEnterpriseId,
    },
    headers: {
      'x-user': row.nm_user ?? 'mirror-worker',
      'X-Mirror-Source': config.sourceName,
    },
    validateStatus: () => true,
    timeout: 15_000,
  });

  if (res.status >= 400) {
    throw new Error(
      `staging /api/production-orders/stop returned ${res.status}: ${JSON.stringify(res.data)}`,
    );
  }

  log.info(
    {
      sourceLogId: row.id_user_logs,
      prodPoId: payload.oldIdProductionOrder,
      stagingPoId: stagingPoId.toString(),
      stopType: payload.stopType,
      status: res.status,
    },
    'replayed order-changed',
  );
}

File: services/mirror-worker-go/internal/replay/order_changed.go
package replay

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"net/http"

	"github.com/jackc/pgx/v5"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/config"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/db"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/translate"
)

// OrderChanged maps to POST /api/production-orders/stop (the operator's
// "change order" action stops the current PO; setup of the new one is a
// separate user_logs row). The DTO ignores the extra payload fields prod
// sends (shouldCreatePo etc.) — class-validator default behavior.
type OrderChangedPayload struct {
	StopType                string      `json:"stopType"`
	Timestamp               string      `json:"timestamp"`
	IDEquipment             int         `json:"idEquipment"`
	OldIDProductionOrder    int64       `json:"oldIdProductionOrder"`
	ProductionOrderQuantity json.Number `json:"productionOrderQuantity"`
}

func OrderChanged(
	cfg *config.Config,
	t *translate.Translator,
	apiToken func() string,
	httpc *http.Client,
	logger *slog.Logger,
) Handler {
	return func(ctx context.Context, _ pgx.Tx, row db.ProdUserLog) error {
		var p OrderChangedPayload
		if err := decodeWithNumbers(row.Payload, &p); err != nil {
			return fmt.Errorf("parse order-changed payload: %w", err)
		}
		if p.IDEquipment == 0 || p.OldIDProductionOrder == 0 || p.StopType == "" {
			return fmt.Errorf("order-changed payload missing fields: %+v", p)
		}
		qty, _ := p.ProductionOrderQuantity.Float64()

		stagingEqID, err := t.Equipment(ctx, p.IDEquipment)
		if err != nil {
			return fmt.Errorf("translate equipment: %w", err)
		}
		stagingPOID, err := t.ProductionOrder(ctx, p.OldIDProductionOrder)
		if err != nil {
			if errors.Is(err, translate.ErrUnmapped) {
				return fmt.Errorf("production_order %d unmapped (no mirror_id_map row and no id_order business-key match — run backfill-pos): %w",
					p.OldIDProductionOrder, err)
			}
			return fmt.Errorf("translate production_order: %w", err)
		}

		body := map[string]any{
			"timestamp":               p.Timestamp,
			"stopType":                p.StopType,
			"idEnterprise":            cfg.StagingEnterpriseID,
			"idEquipment":             stagingEqID,
			"idProductionOrder":       stagingPOID,
			"productionOrderQuantity": qty,
		}
		status, _, err := PostStaging(ctx, cfg, httpc, apiToken(), row,
			"/api/production-orders/stop", body)
		if err != nil {
			return err
		}
		logger.Info("replayed order-changed",
			slog.Int64("sourceLogID", row.IDUserLogs),
			slog.Int64("prodPOID", p.OldIDProductionOrder),
			slog.Int64("stagingPOID", stagingPOID),
			slog.String("stopType", p.StopType),
			slog.Int("status", status))
		return nil
	}
}
